restructuredtext_lint/sphinx.py

```diff
@@ -0,0 +1,154 @@
+    # Extract all possible directives & roles that sphinx is aware of for
+    # all known sphinx domains, and then remove the ones we do not want to
+    # skip (and return them).
+    sp_directives = list(_base_sp_directives)
+    sp_directives.extend(std_domain.StandardDomain.directives)
+    sp_roles = list(_base_sp_roles)
+    sp_roles.extend(std_domain.StandardDomain.roles)
+
+    for (domain, class_name) in [(c_domain, 'CDomain'),
+                                 (cpp_domain, 'CPPDomain'),
+                                 (javascript_domain, 'JavaScriptDomain'),
+                                 (python_domain, 'PythonDomain')]:
+
+        # Get all the domains directives and roles and insert them.
+        domain_class = getattr(domain, class_name)
+        domain_directives = getattr(domain_class, 'directives', [])
+        domain_roles = getattr(domain_class, 'roles', [])
+
+        sp_directives.extend(domain_directives)
+
+        # Ensure that we also use the name prefixed version as well
+        # for example :py:func: and :func: are equivalent and we need to make
+        # sure we register both kinds.
+        sp_directives.extend('%s:%s' % (domain_class.name, item)
+                             for item in domain_directives)
+
+        sp_roles.extend(domain_roles)
+        sp_roles.extend('%s:%s' % (domain_class.name, item)
+                        for item in domain_roles)
+
+    # Filter out the ones we desire to avoid skipping (if any)...
+    if not no_skip_directives:
+        no_skip_directives = []
+    if not no_skip_roles:
+        no_skip_roles = []
+    ok_directives = []
+    for directive in sp_directives:
+        if directive not in no_skip_directives:
+            ok_directives.append(directive)
+    ok_roles = []
+    for role in sp_roles:
+        if role not in no_skip_roles:
+            ok_roles.append(role)
+    return (ok_directives, ok_roles)
+
+
+def register_ignores(no_skip_directives=None, no_skip_roles=None):
+    """Register ignoreable sphinx directives & roles.
+
+    Does nothing if sphinx is not importable.
+
+    :param list/set no_skip_directives: directives to not skip
+    :param list/set no_skip_roles: roles to not skip
+    """
+    if not SPHINX_AVAILABLE:
+        return ([], [])
+    ok_directives, ok_roles = _fetch_roles_directives(
+        no_skip_directives=no_skip_directives,
+        no_skip_roles=no_skip_roles)
+    for directive in ok_directives:
+        rst_directives.register_directive(directive, _IgnoredDirective)
+    for role in ok_roles:
+        rst_roles.register_local_role(role, _ignore_role)
+    return (ok_directives, ok_roles)
+
+
+def unregister_ignores(registered_directives, registered_roles):
+    """Unregister previously registered sphinx directives & roles.
+
+    Does nothing if sphinx is not importable.
+
+    :param list/set registered_directives: directives to unregister
+    :param list/set registered_roles: roles to unregister
```

Comment:
Technically, it is an `iterable` but I think listing it as a `list` type should work fine.